kata-agent: (RFC Implementation) Decoupling Kata Agent and Guest Components I

Apokleos opened the pull request:
Following the great discussions on this RFC, which I believe is key to effectively decoupling the Kata Agent and guest components, I've put together a preliminary implementation of the proposed design.

This is a first pass and will definitely need more fixes and enhancements. My hope is that this code will serve as a concrete example to spark further discussion and inspire more ideas around the RFC.

A huge thanks to Markus @burgerdev  for the excellent design!

**Note: It's still heavily developed.**

It comes from [RFC](https://github.com/kata-containers/kata-containers/issues/11532)

It aims to decouple the initdata process logic from the kata-agent in this PR.

Signed-off-by: Alex Lyn <alex.lyn@antgroup.com>
Signed-off-by: Markus Rudy <mr@edgeless.systems>

Apokleos committed:
kata-agent: Introduce initdata-processor for handling inidata

Adds a dedicated component to locate and parse initdata, validate its
content, and write the resulting configuration to `/run/measured-cfg`.

Signed-off-by: Alex Lyn <alex.lyn@antgroup.com>
Signed-off-by: Markus Rudy <mr@edgeless.systems>

Xynnn007 reviewed (commented):
I like the separate component to do initdata parsing logic. But one thing to notice is that if AA's `bind_init_data` is not executed the config is hard to be said `measured`. We can consider a different name

burgerdev reviewed (commented):
Thanks a lot for driving this, @Apokleos!

mkulke commented on src/agent/initdata-processor/src/initdata.rs:
```diff
@@ -0,0 +1,117 @@
+
+/// It's designed to be run in a separate tokio task to check if a the potential device is the initdata device.
+async fn check_initdata_device(logger: Logger, path: PathBuf) -> Result<Option<String>> {
+    let metadata = match tokio::fs::metadata(&path).await {
+        Ok(m) => m,
+        Err(_) => return Ok(None),
+    };
```
for this you can do the let else pattern, which makes things a bit more legible:

```rust
let Ok(metadata) = tokio::fs::metadata(&path).await else {
   return Ok(None);
}
```

mkulke commented on src/agent/initdata-processor/src/initdata.rs:
```diff
@@ -0,0 +1,117 @@
+    let mut file = match tokio::fs::File::open(&path).await {
+        Ok(f) => f,
+        Err(e) => {
+            error!(
+                logger,
+                "Could not open the potential device `{path:?}`: {e}"
+            );
+            return Ok(None);
```
if there is an error we should not fall back, but fail fast IMO.

mkulke commented on src/agent/initdata-processor/src/initdata.rs:
```diff
@@ -0,0 +1,117 @@
+
+    info!(logger, "Initdata find a potential device: `{path:?}`");
+
+    let mut file = match tokio::fs::File::open(&path).await {
```
see let/else comment above

mkulke commented on src/agent/initdata-processor/src/initdata.rs:
```diff
@@ -0,0 +1,117 @@
+        }
+        _ => {
+            // This covers both the case where magic doesn't match and read errors.
+            // We don't need to bubble up read errors as failures for the whole process.
```
I would argue about this. I think especially for security sensitive logic there should be as little ambiguity as possible: it either works or fails, graceful degradation isn't helpful in this case, but it'll complicate debugging and misconfigurations might go undetected.

mkulke commented on src/agent/initdata-processor/src/initdata.rs:
```diff
@@ -0,0 +1,117 @@
+}
+
+/// Concurrently locate devices using `tokio::spawn`.
+pub async fn locate_device_concurrently(logger: &Logger) -> Result<Option<String>> {
```
this looks pretty manual to me. can we use some self-discovery options like a well-known partition UUID or label or a virtio serial?

e.g.

```
qemu-system-x86_64 -device virtio-blk-pci,drive=initdata.raw,serial=hello-coco
```

will expose the disk like this in the guest:

```
/dev/disk/by-id/virtio-hello-coco
```

burgerdev replied:
I really like this idea, but we should maybe try to keep the diff to the existing implementation minimal (and thus maybe reuse https://github.com/kata-containers/kata-containers/blob/bc75f6a15848cf19c436b3ad7d6bc513c8dbd72d/src/agent/src/initdata.rs#L42

mkulke commented on src/agent/initdata-processor/src/main.rs:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    /// Validates initdata using an external binary.
+    async fn validate_initdata(&self, initdata: &InitData) -> Result<()> {
```
why do we use an external validator command and what is this supposed to do? Maybe it makes sense, but can you add a comment here or to the PR description?

burgerdev replied:
@fitzthum: But separating them is exactly the point of the RFC and the PR at hand!

1. Processing initdata (from annotation to data available in guest) is a feature of the Kata runtime+agent, so should go here. However, since we want to have a strict ordering of initdata processing and agent startup for security reasons, we need it in a separate component.
2. Validating initdata is use-case specific (CoCo uses AA; Contrast and AKS don't; integration tests might not be able to). In order to support the known use cases, we want to make it configurable. I'm not opposed to implementing a default behaviour, but it should still be extensible.

I'd also like to point you to the answers I gave to other reviewers asking similar questions:

https://github.com/kata-containers/kata-containers/pull/11655#issuecomment-3159129045

https://github.com/kata-containers/kata-containers/pull/11655#issuecomment-3160340227

mkulke commented on src/agent/initdata-processor/src/main.rs:
```diff
@@ -0,0 +1,248 @@
+            stdin
+                .write_all(serialized.as_bytes())
+                .context("Failed to write initdata to validator stdin")?;
+        }
```
it should be dropped (and hence closed) at the end of the if-let block

burgerdev commented on src/agent/initdata-processor/src/main.rs:
```diff
@@ -0,0 +1,248 @@
+            .context("Failed to wait for validator completion")?;
+
+        if !output.status.success() {
+            let stderr = String::from_utf8_lossy(&output.stderr);
```
You could also consider `Stdio::inherit` for `stderr`.

Xynnn007 commented:
> the idea is to leave this pluggable

Yes, the reason why I questioned is that now all different platforms share a same API to bind initdata. two ways

1. We could keep it as-is and integrate it into this binary, rather than spawn a new binary/process.
2. release a initdata binding tool (probably dynamically linked for tss-api) in guest-components for each platform and let rootfs builder to select which version to use

burgerdev commented:
> Yes, the reason why I questioned is that now all different platforms share a same API to bind initdata. two ways

This does not really change, does it? I mean, the API before was "pass initdata to AA, move on and see what happens", whereas, with the proposal here, we have a well-defined signal for when initdata validation is done. It's still a unified API across platforms, just different semantics.
 
> 1. We could keep it as-is and integrate it into this binary, rather than spawn a new binary/process.

That goes directly against the stated RFC goal of making initdata validation independent of Kata. If we build it into the initdata processor, everyone who needs slight tweaks (different PCR, use CSP APIs, ...) needs to either also add it here, or carry patches. 

> 2. release a initdata binding tool (probably dynamically linked for tss-api) in guest-components for each platform and let rootfs builder to select which version to use

That's the great thing about making it pluggable: if you want/need a dynamically linked library, you can just add it to the image alongside your validation tool. I think it would be reasonable to provide a default implementation that supports HOSTDATA/MRCONFIGID, probably reusing what's currently in the guest components, but I don't think it makes sense to implement every possible use case in-(Kata)-tree.

Xynnn007 commented:
@burgerdev There is still one thing not clear to me. The processing logic of initdata is divided into two parts. One part is responsible for parsing, and the other part is responsible for binding verification with the TEE platform. 

The latter one is now a crate function in `attester` crate and can be integrated into whatever we want, this is what I mentioned "now all different platforms share a same API to bind initdata".


Therefore, we still have two design approaches for the initdata processing logic. 

1. to separate the binding logic into another (possibly dynamically linked) binary, place it under guest-components, and release it for different archs. showed in this PR.

2. to combine the two into a single binary and release it as a version for different archs.

No matter 1 or 2 the kata rootfs build should index an artifact of gc. Intuitively it would be better to maintain one place than two?

burgerdev commented:
> No matter 1 or 2 the kata rootfs build should index an artifact of gc. Intuitively it would be better to maintain one place than two?

Ah yes, I could have been clearer around that. You're perfectly right that we need to add something to the rootfs regardless. I chose the path of a separate binary so that the _agent itself_ does not need to carry the dependency. Downstreams are more likely to build their own rootfs than their own agent.

Xynnn007 commented:
@burgerdev Oh. Yes, that's fine, we need to decouple this logic from kata-agent. But my question is whether the initdata processing also requires two different components?

mkulke commented:
> And one could also argue that this should be implemented in-tree because the runtime implementation is in-tree, too. But then the question remains: how do downstreams (e.g., but not limited to, peer-pods) implement their bespoke validation? That would then all need to live in the initdata processor.

I might be missing something but does the agent have to be concerned with validation at all?

can we defer that to attestation-agent launch? with a failed snp_report.HOST_DATA == hash(initdata) assertion (or PCR extension for vTPMs) attestation-agent would refuse to launch and hence prohibit the guest from performing attestation?

mkulke commented:
> > I might be missing something but does the agent have to be concerned with validation at all?
> > can we defer that to attestation-agent launch? with a failed snp_report.HOST_DATA == hash(initdata) assertion (or PCR extension for vTPMs) attestation-agent would refuse to launch and hence prohibit the guest from performing attestation?
> 
> The agent - not at all. Do I assume correctly that you meant the initdata processor? In that case:
> 
> The idea is that there's a well-defined handover point, at which the initdata has been validated and services that depend on initdata can be started safely (`initdata.target` in the graphic in #11532). This is particularly relevant for the agent and its policy: if the AA verification is delayed for some reason (or the process just terminates), a malicious host can start any container, including ones that _can perform attestation without AA_.

right, I remember we discussed this situation, thanks for the refresher.

what I don't understand yet: why do we end up launching child processes in the agent again? (the verifier which I assume is equivalent to the initdata "processor"). In the RFC chart I wouldn't see that relationship.

mkulke commented:
> What @Apokleos is implementing here is (mostly) the initdata-processor block from that diagram. The initdata-processor completes before the agent starts, and only the initdata processor shells out to the validator. In a follow-up PR, we would remove the process starting logic (AA, CDH) from the agent and add the equivalent systemd units and a unit for the initdata-processor. Maybe the following sequence illustrates the point better (process lifetime depicted as thick vertical lines):

Ok got it now, that diagram makes it clear, thanks again. I was tripped up by the folder hierarchy, but it probably makes sense to put it in there. now, why does the initdata-processor need to shell out to a validator instead of incorporating validation/binding logic as a library? is there value in using a discrete binary for that?

fitzthum commented:
Catching up here but I will post my thoughts soon.

Xynnn007 commented:
> now, why does the initdata-processor need to shell out to a validator instead of incorporating validation/binding logic as a library? is there value in using a discrete binary for that?

I had the same question and @burgerdev gave an [answer](https://github.com/kata-containers/kata-containers/pull/11655#issuecomment-3159129045). As far as I understand 

1. initdata organization happens in kata-runtime, so it is good to implement parsing logic here.
2. tee specific logic lib would cause a circular dependency from kata to coco. Thus we'd better implement platform specific binding binary tool under gc.

fitzthum reviewed (commented):
The implementation for bare metal coco seems like a bit of an afterthought here.

With the current implementation one thing we had to think about was how to provision configuration to the AA. We have to be careful about the order of operations since the AA is used to check the binding of the init-data, but it also has its own configuration. I can't tell how this approach deals with this kind of situation. It's a little unclear to me what the flow for binding the init-data actually is.

burgerdev commented on src/agent/initdata-processor/src/main.rs:
```diff
@@ -0,0 +1,248 @@
+                    return Err(anyhow::anyhow!("--config-path requires a path argument"));
+                }
+            }
+            "--validator" => {
```
Yes - the idea is that whoever builds a guest image can decide best what binary to put here.

burgerdev commented on src/agent/initdata-processor/src/main.rs:
```diff
@@ -0,0 +1,248 @@
+use kata_types::initdata::InitData;
+
+const MEASURED_CFG_DIR: &str = "/run/measured-cfg";
+const DEFAULT_VALIDATOR: &str = "/usr/bin/initdata-validator";
```
This is discussed in the [RFC](https://github.com/kata-containers/kata-containers/issues/11532). Short answer: not all use cases have an AA, and the current mode of binding is dangerous (cf. https://github.com/kata-containers/kata-containers/pull/11655#issuecomment-3159938988).

burgerdev commented:
> The implementation for bare metal coco seems like a bit of an afterthought here.

On the contrary: this decoupling is specifically designed for bare metal. Maybe the secondary design goal of supporting the current CAA flow (or something close to it) is what's confusing here?

> With the current implementation one thing we had to think about was how to provision configuration to the AA. We have to be careful about the order of operations since the AA is used to check the binding of the init-data, but it also has its own configuration. 

One motivating factor for decoupling was the "cryptographic doom principle": we should avoid using configuration before validating it as far as possible! On the bright side, I don't know of a use case for binding initdata that involves dynamic configuration (as opposed to something fixed in the image), which is why I claim we can separate initdata validation and AA start. This should improve security for everybody, using AA or not.

To be concrete, I know the following ways of binding initdata such that it's reflected in attestation evidence:

* compare initdata hash with `HOSTDATA/MRCONFIGID` and fail on mismatch.
* extend a TPM PCR with the initdata hash and include a TPM quote in attestation.

Do you have other use cases in mind that require dynamic configuration?

> I can't tell how this approach deals with this kind of situation. It's a little unclear to me what the flow for binding the init-data actually is.

Let me try another graphic:

```mermaid
sequenceDiagram
    participant validator
    participant processor
    participant systemd
    participant AA

    systemd->>+processor: start
    processor->>processor: read /dev/initdata
    processor->>+validator: pass initdata
    validator->>validator: bind initdata
    validator->>-processor: success
    processor->>processor: write /run/measured-cfg
    processor->>-systemd: success
    systemd->>systemd: initdata.target completed
    systemd->>+AA: start
    AA ->>AA: read /run/measured-cfg/aa.toml
    AA->>-systemd: 
```

Apokleos committed:
kata-agent: Add Makefile and systemd service for initdata processor

This commit aims to make compile the code easy and introduce
an initdata processor service

Signed-off-by: Alex Lyn <alex.lyn@antgroup.com>
Signed-off-by: Markus Rudy <mr@edgeless.systems>

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

print errors with cause

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

remove tracing

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

remove tokio

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

add unit test for is_initdata_device

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

add unit test for read_initdata

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

make /dev configurable

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

treat udev link like other devices

Apokleos commented on src/agent/initdata-processor/src/initdata.rs:
```diff
@@ -0,0 +1,185 @@
+        let entry = entry?;
+
+        // Just check the file starting with 'vd'
+        if !entry.file_name().to_string_lossy().starts_with("vd") {
```
Yes, copilot and I report that initdata now support both virtio-blk(vdX) and virtio-scsi(sdX), I think we should still support the two ones, WDYT ?

Apokleos commented on src/agent/initdata-processor/src/main.rs:
```diff
@@ -0,0 +1,192 @@
+        initdata_path.add_extension(".json");
+        self.write_file(&initdata_path, &initdata_content)?;
+        self.write_config_files(&initdata)?;
+
```
And one more thing, if I remember correctly that there was some previous discussions about the initdata's suffix being with toml or json, and finally, the toml is chosen.

Apokleos committed:
agent-guests: Try fix issues

(1) Now initdata supports both virtio-blk(vdX) and virtio-scsi(sdX);
(2) The config_path is a directy path we should add its filename with suffix
of TOML(e.g. initdata.toml)
(3) Fix .context("..{e:?}") and drop the placeholder.
(4) Fix systemd units file path with /usr/lib/systemd/system

Signed-off-by: Alex Lyn <alex.lyn@antgroup.com>

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

move initdata directory one level deeper and fix Copilot finding

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

ignore ENOENT when walking device candidates

burgerdev committed:
fixup! kata-agent: Add Makefile and systemd service for initdata processor

run the initdata processor independent of the agent

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

vaildate initdata structurally

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

add license headers

burgerdev committed:
fixup! kata-agent: Add Makefile and systemd service for initdata processor

add license headers

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

format

burgerdev committed:
fixup! kata-agent: Introduce initdata-processor for handling inidata

fix linter warning

burgerdev committed:
osbuilder: integrate initdata-processor

Signed-off-by: Markus Rudy <mr@edgeless.systems>

burgerdev committed:
fixup! osbuilder: integrate initdata-processor

burgerdev committed:
fixup! kata-agent: Add Makefile and systemd service for initdata processor

burgerdev committed:
fixup! kata-agent: Add Makefile and systemd service for initdata processor

Apokleos committed:
fixup! move initdata-processor into workspace

Apokleos committed:
fixup! Enhancements for Makefile and initdata-processor.service

Apokleos committed:
fixup! Add document

Signed-off-by: Alex Lyn <alex.lyn@antgroup.com>

Xynnn007 commented on src/agent/initdata-processor/src/main.rs:
```diff
@@ -0,0 +1,207 @@
+use kata_types::initdata::InitData;
+
+const DEV_DIR: &str = "/dev";
+const MEASURED_CFG_DIR: &str = "/run/measured-cfg";
```
Hi @Apokleos @burgerdev the files extracted from initdata is placed in a new path here compared to [current code](https://github.com/kata-containers/kata-containers/blob/main/src/agent/src/initdata.rs#L25).

Just a FYI and I am ok with either

burgerdev replied:
@Xynnn007: do you know if anything is relying on that path right now? We specified `/run/measured-cfg` for the _extracted values_ of the initdata, but only later I realized that we'd want to place the raw initdata doc somewhere, too. We could place it at `/run/confidential-containers/initdata` for backwards compatibility, wdyt?

Xynnn007 replied:
Hi @burgerdev 

The extracted files are used in
- `Some(AA_CONFIG_PATH)`: https://github.com/kata-containers/kata-containers/blob/main/src/agent/src/main.rs#L500-L510
- `Some(CDH_CONFIG_PATH)`: https://github.com/kata-containers/kata-containers/blob/main/src/agent/src/main.rs#L522-L532
- For agent policy: https://github.com/kata-containers/kata-containers/blob/main/src/agent/src/main.rs#L419-L430

If we put `initdata.toml` under same `/run/confidential-containers/initdata`, there will be risk of collision if a key is named `initdata.toml`. But I am ok with this TO BE HONEST as the initdata binding happens before the file overwritting if a collision really happens.

burgerdev replied:
Sorry, I was confused about what that path meant. Yes, we should not introduce ambiguity between the raw doc and a key `initdata.toml`! That's why I tried writing data to `/run/measured-cfg/initdata/<key>` and raw initdata to `/run/measured-cfg/initdata.toml`, but I did not implement it correctly. I think we can just adapt the paths in the agent once we get to phase 2, and later in the service definitions. As of phase 1, the guest components are still configured by the agent directly and the written data is ignored - that's intentional.

Xynnn007 replied:
Never mind! @burgerdev So let me double check the details with you. Now kata-agent writes initdata under `/run/confidential-containers/initdata`, so

In phase 1: We integrate the initdata processor one-shot systemd service, and writes to `/run/measured-cfg`
In phase 2: add initdata processor target. let kata-agent depend on it
In phase 3: decouple guest-components as systemd services, and use `/run/measured-cfg/initdata/*` instead of `/run/confidential-containers/initdata`. Note that still kata-agent will process initdata and put under `/run/confidential-containers/initdata`, because OPA uses it.
In phase 4: Let OPA use policy under `/run/measured-cfg/initdata/*`

Am I right?

burgerdev replied:
> Never mind! @burgerdev So let me double check the details with you. Now kata-agent writes initdata under `/run/confidential-containers/initdata`, so
> 
> In phase 1: We integrate the initdata processor one-shot systemd service, and writes to `/run/measured-cfg` In phase 2: add initdata processor target. let kata-agent depend on it In phase 3: decouple guest-components as systemd services, and use `/run/measured-cfg/initdata/*` instead of `/run/measured-cfg`. Note that still kata-agent will process initdata and put under `/run/confidential-containers/initdata`, because OPA uses it. In phase 4: Let OPA use policy under `/run/measured-cfg/initdata/*`
> 
> Am I right?

Except of what's probably a typo in phase 3, yes. `instead of /run/measured-cfg ==> instead of /run/confidential-containers/initdata`

burgerdev commented on src/agent/initdata-processor/src/main.rs:
```diff
@@ -0,0 +1,207 @@
+            std::fs::remove_dir_all(&self.config_path)?;
+        }
+
+        let initdata_dir = self.config_path.join("initdata");
+        let initdata_file = initdata_dir.with_extension("toml");
+
+        // Create the config_path.
+        std::fs::create_dir_all(&initdata_dir)
+            .context(format!("Failed to create config path: {:?}", &initdata_dir,))?;
```
Again, the bot is right, I messed that up once more. But considering @Xynnn007's comment in the other thread, we should maybe just go back to (or rather, keep, because I screwed that up) extracting directly into the `config_path` and write the raw doc elsewhere.

burgerdev commented on src/agent/initdata-processor/README.md:
```diff
@@ -0,0 +1,133 @@
+
+### Build Requirements
+
+- Rust toolchain (1.92 or higher)
```
That should indeed align with `rust-toolchain.toml`, maybe we don't need to explicitly document that here at all.

burgerdev commented on src/agent/initdata-processor/src/main.rs:
```diff
@@ -0,0 +1,207 @@
+            "Writing configuration files to: {:?}", self.config_path
+        );
+
+        if std::fs::exists(&self.config_path)? {
+            std::fs::remove_dir_all(&self.config_path)?;
+        }
```
No, that is intentional - the directory is exclusively owned by the initdata processor and we want to make sure that everything in there came from verified initdata (after the processor exits).
